atef/widgets/config/window.py

```diff
@@ -455,379 +455,305 @@ def setup_ui(self):
+        """ Select the TreeItem containing ``data`` """
         for item in walk_tree_items(self.root_item):
-            if item.widget:
-                item.widget.run_check.results_updated.connect(
-                    self.model.data_updated
-                )
+            if item.orig_data is data:
+                self.select_by_item(item)
+                return
 
-        insert_widget(self.tree_view, self.tree_widget)
+    @property
+    def current_item(self) -> Optional[TreeItem]:
+        """ return the currently selected item """
+        sel_model = self.tree_view.selectionModel()
+        try:
+            curr_index = sel_model.selectedIndexes()[0]
+        except IndexError:
+            return None
+        return self.model.data(curr_index, Qt.UserRole)
 
-    def show_selected_page(
+    def show_selected_display(
         self,
         selected: QtCore.QItemSelection,
-        deselected: QtCore.QItemSelection
+        previous: QtCore.QItemSelection
     ) -> None:
-        """
-        Show the proper widget on the right when a tree row is selected.
-
-        For use as a slot connecting to QTreeView.selectionModel().selectionChanged
-
-        This works by hiding the previous widget and showing the new
-        selection, creating the widget object if needed.
+        """ Show selected widget, construct it if necesary """
+        # TODO: show busy cursor?
+        try:
+            current = selected.indexes()[0]
+        except IndexError:
+            logger.debug('no selection made')
+            return
+        item: TreeItem = self.tree_view.model().data(current, Qt.UserRole)
+        logger.debug(f'showing selected display: {item.data(0)}')
+        self.show_page_for_data(item, mode=self.mode)
 
-        Parameters
-        ----------
-        selected : QtCore.QItemSelection
-            Selected region
-        deselected : QtCore.QItemSelection
-            Deselected region (unused)
+    def show_page_for_data(self, item: TreeItem, mode: str = 'edit') -> None:
         """
-        selected_indexes = selected.indexes()
+        Show PageWidget corresponding to ``data`` in ``mode``.
+        If the item's widget is cached, retrieve and show it.  Otherwise the
+        widget must be created.  Schedules oldest widget in cache for deletion
+        if the cache is full.
+        """
+        curr_cache = self.caches[mode]
+        oldest_widget = None
+        if item in curr_cache:
+            new_widget = curr_cache[item]
 
-        item = selected_indexes[0].internalPointer()
+        else:
+            # not in cache, need to build the widget.
+            new_widget = self.create_widget(item, mode)
 
-        if item is None:
-            logger.error('Tree has no currentItem.  Cannot show selected')
-            return
-        if item is self.last_selection:
-            return
+            if len(curr_cache) >= self.max_cache_size:
+                _, oldest_widget = curr_cache.popitem(last=False)
```

Comment:
I worry a bit about this caching mechanism causing unexpected side effects (orphaned widgets, crashes, etc) - but admittedly I'm not the most familiar with the inner workings here. 

Mind commenting on the intention? I assume the concern is memory usage when someone clicks around on a bunch of widgets and loads them all up?

Reply:
Memory usage was the concern, since it's easy to just hold the down-arrow and scroll through all the widgets.  (or spam the next-button in run-mode)

I haven't run into issues yet, and the shift to making each PageWidget self sufficient is important to enabling this.  (as in a PageWidget will never ask for methods on other PageWidgets).  If a PageWidget is deleted, nothing should ever want to access it.  

The concerns are valid though, I can't confidently say there will be no crashes.  (as your previous comment shows)

Reply:
Do you know how much memory, say, ~1000 (hidden) page widgets uses? It would be useful to quantify this to understand what the real impacts are here. I personally suspect that crashing concerns on this PR and on any future expansions to atef are much more real than the RAM usage invested here, but without knowing the actual numbers it's impossible to confirm.

Reply:
our server has 50.46 GB of RAM
opening the big file with no pages loaded is 0.8% memory usage = 0.4 GB consumed
with all the comparison pages loaded we get 3.5% usage -> 2.7% net = 1.362 GB RAM for just the pages
the config has 2541 Equals comparisons
so this means we eat up approx 536 kB per comparison page

I _think_ this means we're typically going to be OK because the example here is not reasonable or maintainable